DigitalLearningSolutions.Web/ControllerHelpers/DateValidator.cs

```diff
@@ -1,172 +1,135 @@
-                emptyElements.Add("month");
-            }
-
-            if (year == 0)
-            {
-                emptyElements.Add("year");
-            }
-
-            if (emptyElements.Any())
-            {
-                return name + " must include a " + string.Join(" and ", emptyElements);
-            }
-
-            return name + " must be a real date";
-        }
-    }
-}
+﻿namespace DigitalLearningSolutions.Web.ControllerHelpers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel.DataAnnotations;
+
+    public static class DateValidator
+    {
+        public static DateValidationResult ValidateDate(
+            int? day,
+            int? month,
+            int? year,
+            string name = "Date",
+            bool required = false
+        )
+        {
+            if (!day.HasValue && !month.HasValue && !year.HasValue)
+            {
+                return required ? new DateValidationResult(name + " is required") : new DateValidationResult();
+            }
+
+            if (!day.HasValue || !month.HasValue || !year.HasValue)
+            {
+                var errorMessage = GetMissingValuesErrorMessage(day, month, year, name);
+                return new DateValidationResult(!day.HasValue, !month.HasValue, !year.HasValue, errorMessage);
+            }
+
+            return ValidateDate(day.Value, month.Value, year.Value, name);
+        }
+
+        private static DateValidationResult ValidateDate(int day, int month, int year, string name)
+        {
+            var invalidDay = day < 1 || day > 31;
+            var invalidMonth = month < 1 || month > 12;
+            var invalidYear = year < 1;
+
+            if (invalidDay || invalidMonth || invalidYear)
+            {
+                return new DateValidationResult(invalidDay, invalidMonth, invalidYear, name + " must be a real date");
+            }
+
+            try
+            {
+                var date = new DateTime(year, month, day);
+                if (date <= DateTime.Today)
+                {
+                    return new DateValidationResult(name + " must be in the future");
+                }
+            }
+            catch (ArgumentOutOfRangeException)
+            {
+                return new DateValidationResult(name + " must be a real date");
+            }
+
+            return new DateValidationResult();
+        }
+
+        private static string GetMissingValuesErrorMessage(int? day, int? month, int? year, string name)
+        {
+            var missingValues = new List<string>();
+            if (!day.HasValue)
+            {
+                missingValues.Add("day");
+            }
+
+            if (!month.HasValue)
+            {
+                missingValues.Add("month");
+            }
+
+            if (!year.HasValue)
+            {
+                missingValues.Add("year");
+            }
+
+            return name + " must have a " + string.Join(" and ", missingValues);

```

Comment:
Do we want the string.Join() separator " and " to instead be " and a " to be consistent with the earlier part of the sentence?